File: cpp/ExpirableProduct.cpp
// cpp/ExpirableProduct.cpp
#include "ExpirableProduct.h"

ExpirableProduct::ExpirableProduct(std::string name, double price, int quantity, std::chrono::system_clock::time_point expiryDate)
    : Product(name, price, quantity), expiryDate(expiryDate) {}

std::chrono::system_clock::time_point ExpirableProduct::getExpiryDate() const {
    return expiryDate;
}

bool ExpirableProduct::isExpired() const {
    return std::chrono::system_clock::now() > expiryDate;
}

File: csharp/ExpirableProduct.cs
// csharp/ExpirableProduct.cs
using System;

namespace ECommerce
{
    public abstract class ExpirableProduct : Product
    {
        public DateTime ExpiryDate { get; protected set; }

        public ExpirableProduct(string name, double price, int quantity, DateTime expiryDate)
            : base(name, price, quantity)
        {
            ExpiryDate = expiryDate;
        }

        public override bool IsExpired()
        {
            return DateTime.Now > ExpiryDate;
        }
    }
}
